lib/semantics/expression.cc

```diff
@@ -1762,21 +1773,26 @@ auto ExpressionAnalyzer::GetCalleeAndArguments(
       return CalleeAndArguments{
           ProcedureDesignator{std::move(specificCall->specificIntrinsic)},
           std::move(specificCall->arguments)};
-    } else {
-      return std::nullopt;
     }
   } else {
     CheckForBadRecursion(name.source, ultimate);
     if (ultimate.has<semantics::GenericDetails>()) {
-      symbol = ResolveGeneric(*symbol, arguments);
+      symbol = ResolveGeneric(
+          *symbol, arguments, std::nullopt, mightBeStructureConstructor);
     }
     if (symbol) {
-      return CalleeAndArguments{
-          ProcedureDesignator{*symbol}, std::move(arguments)};
-    } else {
-      return std::nullopt;
+      if (symbol->GetUltimate().has<semantics::DerivedTypeDetails>()) {
+        if (mightBeStructureConstructor) {
+          return CalleeAndArguments{
+              semantics::SymbolRef{*symbol}, std::move(arguments)};
+        }
+      } else {
+        return CalleeAndArguments{
+            ProcedureDesignator{*symbol}, std::move(arguments)};
+      }
```

Comment:
You could write this as:
```
if (mightBeStructureConstructor || !symbol->GetUltimate().has<semantics::DerivedTypeDetails>()) {
  return CalleeAndArguments{semantics::SymbolRef{*symbol}, std::move(arguments)};
}
```
This avoids the duplicate constructor call but does make the condition more complicated.

Reply:
The constructor calls are actually not identical -- one is a `SymbolRef` (signifying a structure constructor) and the other is a `ProcedureDesignator`.